File: src/main/extractor/OCRExtractor/ScreenCapturer.ts
import type { DModel as M } from 'win32-def';
import { promisify } from 'util';
import sharp from 'sharp';
import { gdi32, user32 } from '@main/win32';

export class ScreenCapturer {
    private hwnd?: M.HWND;
    constructor(
        public gamePids: number[],
        hwnd?: M.HWND
    ) {
        if (hwnd) this.hwnd = hwnd;
        else this.findWindow();
    }

    private findWindow() {
        const ret = user32.EnumWindows(hwnd => {
            const lpdwProcessId = Buffer.alloc(4, 0);
            user32.GetWindowThreadProcessId(hwnd, lpdwProcessId);
            const pid = lpdwProcessId.readUInt32LE(0);
            if (user32.IsWindowEnabled(hwnd) && user32.IsWindowVisible(hwnd) && this.gamePids.includes(pid)) {
                this.hwnd = hwnd;
                return 0;
            } else {
                return 1;
            }
        }, 0);
        delete ret.callbackEntry;
    }

    public async capture() {
        if (!this.hwnd) this.findWindow();
        if (!this.hwnd) return sharp(Buffer.from(''));

        const rect = Buffer.alloc(4 * 4, 0);
        user32.GetWindowRect(this.hwnd, rect);
        const left = rect.readInt32LE(0);
        const top = rect.readInt32LE(4);
        const right = rect.readInt32LE(8);
        const bottom = rect.readInt32LE(12);
        const width = right - left;
        const height = bottom - top;

        const hwndDC = user32.GetWindowDC(this.hwnd);
        const saveDC = gdi32.CreateCompatibleDC(hwndDC);
        const saveBitmap = gdi32.CreateCompatibleBitmap(hwndDC, width, height);
        gdi32.SelectObject(saveDC, saveBitmap);
        await promisify(user32.PrintWindow.async)(this.hwnd, saveDC, 0x00000001 | 0x00000002 /* PW_CLIENTONLY | PW_RENDERFULLCONTENT */);
        const buf = Buffer.alloc(width * height * 4, 0);
        promisify(gdi32.GetBitmapBits.async)(saveBitmap, width * height * 4, buf);
        gdi32.DeleteObject(saveBitmap);
        user32.ReleaseDC(this.hwnd, hwndDC);
        gdi32.DeleteObject(saveBitmap);

        const img = sharp(buf, { raw: { width, height, channels: 4 } });
        const [rBuffer, gBuffer, bBuffer, aBuffer] = await Promise.all([
            img.clone().extractChannel(2).raw().toColorspace('b-w').toBuffer(),
            img.clone().extractChannel(1).raw().toColorspace('b-w').toBuffer(),
            img.clone().extractChannel(0).raw().toColorspace('b-w').toBuffer(),
            img.clone().extractChannel(3).raw().toColorspace('b-w').toBuffer()
        ]);
        const r = sharp(rBuffer, { raw: { width, height, channels: 1 } });
        const rgba = r
            .joinChannel(gBuffer, { raw: { width, height, channels: 1 } })
            .joinChannel(bBuffer, { raw: { width, height, channels: 1 } })
            .joinChannel(aBuffer, { raw: { width, height, channels: 1 } });
        return sharp(await rgba.raw().toBuffer(), { raw: { width, height, channels: 4 } });
    }

    public destroy() { }
}

File: src-tauri/src/win32/screen_capturer.rs
//! Screen capture — replaces `native/addons/ScreenCapturer`.

use windows::core::BOOL;
use windows::Win32::Foundation::{HWND, LPARAM, RECT};
use windows::Win32::Graphics::Dwm::{DwmGetWindowAttribute, DWMWA_EXTENDED_FRAME_BOUNDS};
use windows::Win32::Graphics::Gdi::{
    CreateCompatibleBitmap, CreateCompatibleDC, DeleteDC, DeleteObject, GetDIBits, GetWindowDC,
    ReleaseDC, SelectObject, BITMAPINFO, BITMAPINFOHEADER, BI_RGB, DIB_RGB_COLORS,
};
use windows::Win32::Storage::Xps::{PrintWindow, PRINT_WINDOW_FLAGS};
use windows::Win32::UI::Input::KeyboardAndMouse::IsWindowEnabled;
use windows::Win32::UI::WindowsAndMessaging::{
    EnumWindows, GetWindowThreadProcessId, IsWindowVisible, PW_RENDERFULLCONTENT,
};

/// Find the first visible, enabled window belonging to any of the given PIDs.
pub fn find_window(pids: &[u32]) -> u64 {
    struct Data {
        pids: Vec<u32>,
        result: HWND,
    }

    unsafe extern "system" fn enum_proc(hwnd: HWND, lparam: LPARAM) -> BOOL {
        let data = &mut *(lparam.0 as *mut Data);
        let mut pid: u32 = 0;
        GetWindowThreadProcessId(hwnd, Some(&mut pid));
        if data.pids.contains(&pid)
            && IsWindowEnabled(hwnd).as_bool()
            && IsWindowVisible(hwnd).as_bool()
        {
            data.result = hwnd;
            return BOOL(0);
        }
        BOOL(1)
    }

    let mut data = Data {
        pids: pids.to_vec(),
        result: HWND::default(),
    };
    unsafe {
        let _ = EnumWindows(Some(enum_proc), LPARAM(&mut data as *mut Data as isize));
    }
    data.result.0 as u64
}

/// Check that a window handle still belongs to one of the given PIDs.
///
/// After the game window is destroyed, Windows may reuse the handle value for
/// an unrelated window (desktop, browser, …); `capture()` would then silently
/// succeed on the wrong window and keep triggering OCR. Callers use this to
/// detect that the tracked window is gone.
pub fn window_belongs_to(hwnd_value: u64, pids: &[u32]) -> bool {
    if hwnd_value == 0 {
        return false;
    }
    let hwnd = HWND(hwnd_value as *mut _);
    let mut pid: u32 = 0;
    unsafe {
        GetWindowThreadProcessId(hwnd, Some(&mut pid));
    }
    pid != 0 && pids.contains(&pid)
}

/// Captured window image.
#[derive(Clone)]
pub struct CapturedImage {
    pub width: u32,
    pub height: u32,
    pub buffer: Vec<u8>,
}

/// Capture a window's content via PrintWindow + GetDIBits.
pub fn capture(hwnd_value: u64) -> Result<CapturedImage, String> {
    let hwnd = HWND(hwnd_value as *mut _);

    unsafe {
        let mut rect = RECT::default();
        DwmGetWindowAttribute(
            hwnd,
            DWMWA_EXTENDED_FRAME_BOUNDS,
            &mut rect as *mut RECT as *mut _,
            std::mem::size_of::<RECT>() as u32,
        )
        .map_err(|e| format!("DwmGetWindowAttribute failed: {e}"))?;

        let width = rect.right - rect.left;
        let height = rect.bottom - rect.top;
        if width <= 0 || height <= 0 {
            return Err("Invalid window size".into());
        }

        let buffer_size = (width as usize) * (height as usize) * 4;
        let mut buffer = vec![0u8; buffer_size];

        let hwnd_dc = GetWindowDC(Some(hwnd));
        let save_dc = CreateCompatibleDC(Some(hwnd_dc));
        let save_bitmap = CreateCompatibleBitmap(hwnd_dc, width, height);
        let _old = SelectObject(save_dc, save_bitmap.into());
        // 必须带 PW_RENDERFULLCONTENT：D3D/DXGI 硬件加速窗口只有在该标志下
        // 才会把渲染内容画进 PrintWindow 的 DC，否则截出来是白屏。
        let _ = PrintWindow(hwnd, save_dc, PRINT_WINDOW_FLAGS(PW_RENDERFULLCONTENT));

        let bi = BITMAPINFO {
            bmiHeader: BITMAPINFOHEADER {
                biSize: std::mem::size_of::<BITMAPINFOHEADER>() as u32,
                biWidth: width,
                biHeight: height,
                biPlanes: 1,
                biBitCount: 32,
                biCompression: BI_RGB.0,
                biSizeImage: 0,
                biXPelsPerMeter: 0,
                biYPelsPerMeter: 0,
                biClrUsed: 0,
                biClrImportant: 0,
            },
            ..Default::default()
        };

        GetDIBits(
            hwnd_dc,
            save_bitmap,
            0,
            height as u32,
            Some(buffer.as_mut_ptr() as *mut _),
            &bi as *const BITMAPINFO as *mut _,
            DIB_RGB_COLORS,
        );

        let _ = DeleteObject(save_bitmap.into());
        let _ = DeleteDC(save_dc);
        let _ = ReleaseDC(Some(hwnd), hwnd_dc);

        Ok(CapturedImage {
            width: width as u32,
            height: height as u32,
            buffer,
        })
    }
}
